src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -89,25 +89,28 @@ private static void ReadCollectionItem(ODataReader reader, Stack<ODataItemBase>
                     }
                     else
                     {
-                        ODataItemBase parentItem = itemsStack.Peek();
-                        ODataResourceSetWrapper parentResourceSet = parentItem as ODataResourceSetWrapper;
-                        if (parentResourceSet != null)
-                        {
-                            parentResourceSet.Resources.Add(resourceWrapper);
-                        }
-                        else
-                        {
-                            ODataNestedResourceInfoWrapper parentNestedResource = (ODataNestedResourceInfoWrapper)parentItem;
-                            Contract.Assert(parentNestedResource.NestedResourceInfo.IsCollection == false, "Only singleton nested properties can contain resource as their child.");
-                            Contract.Assert(parentNestedResource.NestedItems.Count == 0, "Each nested property can contain only one resource as its direct child.");
-                            parentNestedResource.NestedItems.Add(resourceWrapper);
-                        }
+                        AddResourceToParent(itemsStack, resourceWrapper);
                     }
 
                     itemsStack.Push(resourceWrapper);
                     break;
 
+                case ODataReaderState.DeletedResourceStart:
+                    ODataDeletedResource deletedResource = (ODataDeletedResource)reader.Item;
+                    Contract.Assert(deletedResource != null, "Deleted resource should not be null");
```

Comment:
Given that the statement above this one does a cast, i.e., `(ODataDeletedResource)reader.Item` and not type conversion, i.e., `reader.Item as ODataDeletedResource`, which scenarios would give rised to `deletedResource` variable being null?

Reply:
The reader state just tells us that we are currently reading an `ODataDeletedResource`. The item could be null or the cast could lead to a null. no? I don't think the case statement guarantees that we have the `ODataDeletedResource`. I could be wrong here.

Reply:
can reader.Item contain an exception or something like that?

Reply:
all the case statements have these asserts..can i leave them just for uniformity or I remove all of them even in the other cases?